internal_transfers/actions/tests/database.spec.ts

```
@@ -151,10 +154,128 @@ describe("test database insertion methods", () => {
+      {
+        token: token1,
+        amount:
+          115792089237316195423570985008687907853269984665640564039457584007913129639935n,
+      },
+      { token: token2, amount: 5678n },
+    ];
+    const blockNumber = 1;
+    const gasUsed = 2;
+    const simulationID = "sim-id";
+    const largeBigInt = 99999999999999999999999999999999999999999999999999n;
+    const exampleLog = {
+      address: ethers.ZeroAddress,
+      // The indexed topics from the event log
+      topics: [ethers.ZeroHash, ethers.ZeroHash],
+      data: ethers.ZeroHash,
+    };
+    const exampleEthDelta = new Map([["0x2", largeBigInt]]);
+    const simulationDatum: SettlementSimulationData = {
+      txHash,
+      full: {
+        simulationID,
+        gasUsed,
+        blockNumber,
+        // Note there are 2 logs here!
+        logs: [exampleLog, exampleLog],
+        ethDelta: exampleEthDelta,
+      },
+      reduced: {
+        simulationID,
+        gasUsed,
+        blockNumber,
+        // and only 1 log here!
+        logs: [exampleLog],
+        ethDelta: exampleEthDelta,
+      },
+      winningSettlement: {
+        solver: ethers.ZeroAddress,
+        simulationBlock: blockNumber,
+        reducedCallData: "0xca11da7a",
+        fullCallData: "0xca11da7a000000",
+      },
+    };
+    // insertion works.
+    await insertPipelineResults(
+      db,
+      simulationDatum,
+      imbalances,
+      eventMeta,
+      settlementEvent
+    );
+    const recordCountQuery = sql`select count(*)
+                from internalized_imbalances i
+                       join settlements s
+                            on i.tx_hash = s.tx_hash
+                       join settlement_simulations ss
+                            on i.tx_hash = ss.tx_hash`;
+    let results = await db.query(recordCountQuery);
+    expect(results.length).toEqual(1);
+    expect(results[0]).toEqual({ count: 2n });
+
+    // Idempotency & insertion works ONLY IF ALL three insertions pass!
```

My understanding is that the function `insertPipelineResults` isn't idempotent because if you run it again with the same input it errors instead of succeeding. Am I missing something?

---

Its idempotent in the sense that it cannot be applied twice. Previously it was actually idempotent, but then it would essentially perform nothing in silence in situations when it should at least log that the operation had already been performed. I can use a different name here, but I didn't think it made sense to loose this information by "swallowing it".

---

> Its idempotent in the sense that it cannot be applied twice.

An idempotent function can be applied twice and this would be the same as applying it once. Strictly speaking, even the extra logging is too much. To me, the word "idempotent" here is confusing and I would be using this function wrong if I was relying on what I think is idempotency.

---

That is reasonable. So then I will remove the term "idempotent" here unless you think the operation should actually just act that way. 

For program efficiency, my opinion is that its better to check if an operation has already been performed (to avoid redundant simulation costs etc...), so what will actually become idempotent is the method which calls this and implements some sort of try/catch to handle the insertion failure. 
I just wanted to make sure we were informed of such situations instead of failing silently.

---

Removed the term.